File: PeroLauncher/src/everything.rs
use std::path::PathBuf;
use std::fs;
use std::io::{self, Cursor};
use tauri::{AppHandle, Emitter};
use zip::ZipArchive;
use reqwest::blocking::Client;

use tauri::Manager;

pub fn get_es_dir(app: &AppHandle) -> PathBuf {
    // 1. 优先探测开发环境
    let dev_root = crate::get_workspace_root();
    let dev_path = dev_root.join("backend/nit_core/tools/core/FileSearch");
    if dev_path.exists() {
        return dev_path;
    }

    // 2. 释放环境 (Release)
    if let Ok(resource_dir) = app.path().resource_dir() {
        let pkg_path = resource_dir.join("backend/nit_core/tools/core/FileSearch");
        if pkg_path.exists() {
            return pkg_path;
        }
        
        let flat_path = resource_dir.join("nit_core/tools/core/FileSearch");
        if flat_path.exists() {
            return flat_path;
        }
    }

    dev_path
}

pub fn emit_log(app: &AppHandle, msg: String) {
    let _ = app.emit("es-log", msg);
}

pub fn check_es_installed(app: &AppHandle) -> bool {
    let dir = get_es_dir(app);
    let exe = dir.join("es.exe");
    exe.exists()
}

pub fn install_es(app: AppHandle) -> Result<(), String> {
    let dir = get_es_dir(&app);
    emit_log(&app, format!("Checking ES tool in: {:?}", dir));

    if check_es_installed(&app) {
        emit_log(&app, "ES tool already installed.".to_string());
        return Ok(());
    }

    emit_log(&app, "ES tool not found. Starting download...".to_string());
    fs::create_dir_all(&dir).map_err(|e| e.to_string())?;

    // URL for ES command line tool (x64)
    let url = "https://www.voidtools.com/ES-1.1.0.27.x64.zip";
    
    let client = Client::new();
    let response = client.get(url).send().map_err(|e| format!("Download failed: {}", e))?;
    
    if !response.status().is_success() {
         return Err(format!("Download failed with status: {}", response.status()));
    }

    // Read bytes
    let bytes = response.bytes().map_err(|e| e.to_string())?;
    let reader = Cursor::new(bytes);
    let mut archive = ZipArchive::new(reader).map_err(|e| e.to_string())?;

    // Extract only es.exe
    let mut found = false;
    for i in 0..archive.len() {
        let mut file = archive.by_index(i).map_err(|e| e.to_string())?;
        if file.name() == "es.exe" {
            let outpath = dir.join("es.exe");
            let mut outfile = fs::File::create(&outpath).map_err(|e| e.to_string())?;
            io::copy(&mut file, &mut outfile).map_err(|e| e.to_string())?;
            found = true;
            break;
        }
    }

    if !found {
        return Err("es.exe not found in downloaded zip".to_string());
    }
    
    emit_log(&app, "ES tool installation complete.".to_string());
    Ok(())
}

File: electron/main/services/everything.ts
import { BrowserWindow } from 'electron'
import path from 'path'
import fs from 'fs-extra'
import AdmZip from 'adm-zip'
import axios from 'axios'
import { join } from 'path'

const ES_URL = "https://www.voidtools.com/ES-1.1.0.27.x64.zip"

function getWorkspaceRoot() {
    return process.cwd()
}

export function getEsDir() {
    // 1. Dev
    // 1. 开发环境
    const devRoot = getWorkspaceRoot()
    const devPath = join(devRoot, 'backend/nit_core/tools/core/FileSearch')
    if (fs.existsSync(devPath)) return devPath
    
    // 2. Prod (resources)
    // 2. 生产环境 (resources)
    const resourcePath = process.resourcesPath
    const pkgPath = join(resourcePath, 'backend/nit_core/tools/core/FileSearch')
    if (fs.existsSync(pkgPath)) return pkgPath
    
    // Fallback
    // 后备方案
    return devPath
}

export function checkEsInstalled() {
    const dir = getEsDir()
    const exe = join(dir, 'es.exe')
    return fs.existsSync(exe)
}

export async function installEs(window: BrowserWindow) {
    const dir = getEsDir()
    const emit = (msg: string) => window.webContents.send('es-log', msg)
    
    emit(`Checking ES tool in: ${dir}`)
    
    if (checkEsInstalled()) {
        emit("ES tool already installed.")
        return true
    }
    
    emit("ES tool not found. Starting download...")
    await fs.ensureDir(dir)
    
    try {
        const response = await axios({
            method: 'get',
            url: ES_URL,
            responseType: 'arraybuffer'
        })
        
        const zipBuffer = Buffer.from(response.data)
        const zip = new AdmZip(zipBuffer)
        const zipEntries = zip.getEntries()
        
        let found = false
        for (const entry of zipEntries) {
            if (entry.entryName === 'es.exe') {
                zip.extractEntryTo(entry, dir, false, true)
                found = true
                break
            }
        }
        
        if (!found) {
            throw new Error("es.exe not found in downloaded zip")
        }
        
        emit("ES tool installation complete.")
        return true
    } catch (e: any) {
        emit(`Download/Install failed: ${e.message}`)
        return false
    }
}
